### ISSUE TITLE
Create an updated Free-free model

### ISSUE THREAD

**Issue by zonca** [galsci/pysm#197]
@brandonshensley @seclark @giuspugl the `f1` model we inherited from PySM 2 has sharp features in the template, and causes complications with spherical harmonics transforms.

For example, if we evaluate the model at 20GHz, then we want to upscale it to 2048 using spherical harmonics instead of `ud_grade`, we get significant ringing.

![image](https://github.com/user-attachments/assets/814757d6-fb85-4954-ba4a-9af78b59bee6)
![image](https://github.com/user-attachments/assets/44349579-da30-47dd-8eaa-b3321797e027)

See notebook here: https://gist.github.com/zonca/c44d70c72b211ca6b9a1b999f68b36c0

The workaround for now is to use `ud_grade`, which creates some strange features in the spectra at high ell but it is otherwise well-behaved. I don't think it is worth investing time trying to solve this issue.

I think we should prioritize creating an updated Free-free model, even if there are no new measurements, we could use the `d10`/`s5`-style high ell extension. Maybe someone in the Panexp group could volunteer to do this, I can help reviewing the code.

**Comment by seclark** [comment]
Sounds like a good plan to me! Let's bring this up at the next PanEx PySM chat. Maybe you can show this. With the paper winding down I bet we'll have some free energy available.

**Comment by zonca** [comment]
# Discussion on Panexp  call 9 April 2025

We spent 20 min talking about this in the Panexp call today.
The plan is to create a `f2` model using as template the dataset published here:

https://ui.adsabs.harvard.edu/abs/2020A%26A...633A.150H/abstract

## Adding small scales

The plan is to do something similar to what we did for `d10` and `s5`, see this notebook in the PySM docs:

https://pysm3.readthedocs.io/en/latest/preprocess-templates/synchrotron_template_logpoltens.html

of course we do not need the logpoltens formalism given this is T only.

We want to extrapolate the slope of the small scales the same way, see the plot in cell 31:

![Image](https://github.com/user-attachments/assets/80cd1799-cee0-4386-9a0b-3f3ddf904164)

Then want to modulate it: https://pysm3.readthedocs.io/en/latest/preprocess-templates/synchrotron_template_logpoltens.html#(new)-modulation-for-intensity-and-polarization

## Plan

1) We would like to have a notebook that compares this new template to the template used in `f1` (which is at 30 GHz https://portal.nersc.gov/cfs/cmb/pysm-data/pysm_2/ff_t_new.fits), so that people know what to expect.
2) Produce a notebook that adds small scales and creates a nside 2048 map, does some plots inspecting them
3) After we have reviewed the notebook, add at the bottom of that notebook the capability of generating a map with ellmax of 8192*2 with a fixed seed and then save maps at Nside 2048, 4096 and 8192, use the same technique as https://github.com/galsci/pysm/blob/main/docs/preprocess-templates/utils_synch_generate_map.ipynb

**Comment by brandonshensley** [comment]
A small note that there even more up to date templates from the same collaboration: https://zenodo.org/records/10523170 (paper: https://ui.adsabs.harvard.edu/abs/2024A%26A...690A.314H/abstract)

**Comment by giuspugl** [comment]
Hello 
that 's exactly what i had in mind ! thanks a lot ! Seems we are gonna have tons of fun with free-free!

**Comment by fazlurahman** [comment]
@zonca @giuspugl Is there a place where I can find the PySM2 free-free template without small-scale realizations added? I can certainly generate one from the Planck Commander EM maps, but it would be great if there’s a ready-made version available.

Thanks.

**Comment by fazlurahman** [comment]
Also, would be great to know how you did convert from EM to brightness temperature. I find this formula in Dickinson (2003), but I’d like to make sure I’m using the correct one.

![Image](https://github.com/user-attachments/assets/f12194a7-7f56-42f9-994b-84f7233c104f)

**Comment by brandonshensley** [comment]
> Also, would be great to know how you did convert from EM to brightness temperature. I find this formula in Dickinson (2003), but I’d like to make sure I’m using the correct one.
> 
> ![Image](https://github.com/user-attachments/assets/f12194a7-7f56-42f9-994b-84f7233c104f)

The exact procedure is documented in Section 2.4 of the original [PySM paper](https://ui.adsabs.harvard.edu/abs/2017MNRAS.469.2821T/abstract). See Table 4 [here](https://ui.adsabs.harvard.edu/abs/2016A%26A...594A..10P/abstract) for manipulating the Commander parameter maps.

**Comment by fazlurahman** [comment]
Thanks @brandonshensley. Yeah, I was looking at the original PySM paper, but it points to Draine (2011) for the analytic formula, which has too many equations; easy to miss numerical factors. 

This table in the Planck paper is quite helpful. 

![Image](https://github.com/user-attachments/assets/184efb6a-38ac-41a8-99b0-d8ab7bcd6d81)

**Comment by zonca** [comment]
@fazlurahman it would be useful to track also the ongoing work, not only the final version of the notebook.
So I suggest you put the current **executed** version of the notebook on gist (I like to use `gh gist create`, but you can also upload manually to gist.github.com) and put a link here, so we have a track of what was done. Thanks.

**Comment by fazlurahman** [comment]
@zonca. Sure, I’ll post the notebook on Gist soon. Just got a bit busy with other things.

**Comment by fazlurahman** [comment]
Hi all,

Sorry for the delay in getting back to this. Here’s the notebook I presented during the call:

https://gist.github.com/fazlurahman/87345c231de451d9da84d899d831ad0d

I haven’t yet incorporated the comments and suggestions from our discussion, but I’ll update the notebook once I do. Also, apologies for the messy and somewhat disorganized style!

**Comment by fazlurahman** [comment]
Uploaded a clean version of a standalone notebook comparing Planck and HE EM maps, along with the free-free intensity maps.

https://gist.github.com/fazlurahman/87345c231de451d9da84d899d831ad0d#file-he_planck_comparison-ipynb

**Comment by fazlurahman** [comment]
Quick question: I noticed that the formalism implemented in the notebook (that @zonca shared above) uses different values for parameters like the smoothing scale, patch radius, and the sigmoid function compared to those mentioned in the paper. I’m wondering if there’s another version of the notebook for synchrotron small-scale generation that follows the parameter choices given in the paper. For instance, the radius of the circular mask in the notebook is 15.6 while the one in the paper is 11.3.

**Comment by fazlurahman** [comment]
I have several questions about the parameter choices in the paper and notebook, which I assume are specifically tuned for synchrotron. I’m currently adjusting them for the free-free maps to try and get a better modulation map.

I have listed them here. 

![Image](https://github.com/user-attachments/assets/18b1c54d-d463-40cf-92de-c430ecf4ad2e)

**Comment by fazlurahman** [comment]
It would be great if I could clarify my queries over a Zoom call.

**Comment by seclark** [comment]
@giuspugl could you take a look? and @fazlurahman are you available to join the call tomorrow? We'll focus on the referee report but should also clear this up...

**Comment by fazlurahman** [comment]
@seclark I’ll be joining the call tomorrow and am happy to discuss then. However, many of my questions and comments are quite minor or technical, and I’m not sure they’ll be relevant for everyone on the call. It might be better to have a separate short chat with @giuspugl or with you all at a later time.

**Comment by casasjm** [comment]
I just took a look at the notebook. Great work! It might be useful to show the difference map and the residual power spectra between what you found and the Commander results. What do you think?

I'm not sure about the parameter choices you said. They surely know about it and say you something!

**Comment by fazlurahman** [comment]
Posting (Planck EM - HE EM) map for a quick check.  

![Image](https://github.com/user-attachments/assets/0ab2a453-cf62-40ab-b33f-75179418bd7a)

**Comment by giuspugl** [comment]
this is very helpful thanks! and i think it made me to favor the ff map from HE than the Planck one, the presence of so many point sources (surely extragalactic) makes me tending towards this. how does this plane sound ?

**Comment by fazlurahman** [comment]
You are right, @giuspugl. In fact, HE paper highlights point source handling as a key improvement in their analysis.

**Comment by fazlurahman** [comment]
A preliminary version of the free-free map with added small-scale features. Sharing here as an update on the ongoing work.

![Image](https://github.com/user-attachments/assets/d1ba1fcb-4f4e-4122-bd40-c28f21f10561)

**Comment by fazlurahman** [comment]
Here are the power spectra of the map at various stages. The issue is that after modulation, the power gets suppressed. I’ll tweak the parameters to see what works best.

![Image](https://github.com/user-attachments/assets/8e9b274a-17e8-4e2d-9094-5a8d65bdf432)

**Comment by giuspugl** [comment]
there should be something wrong, why the total Cls don't match with the large scale ones?

**Comment by zonca** [comment]
> Quick question: I noticed that the formalism implemented in the notebook (that [@zonca](https://github.com/zonca) shared above) uses different values for parameters like the smoothing scale, patch radius, and the sigmoid function compared to those mentioned in the paper. I’m wondering if there’s another version of the notebook for synchrotron small-scale generation that follows the parameter choices given in the paper. For instance, the radius of the circular mask in the notebook is 15.6 while the one in the paper is 11.3.

the notebooks in the pysm docs are the ones actually used to generate the templates inside PySM, so they should be the primary reference.

**Comment by zonca** [comment]
> A preliminary version of the free-free map with added small-scale features. Sharing here as an update on the ongoing work.
> 
> You can see the ringing due to the sigmoid function filtering (as discussed in the last call, this ringing is less prominent in the logpoltens space for synchrotron and dust).
> 
> ![Image](https://github.com/user-attachments/assets/d1ba1fcb-4f4e-4122-bd40-c28f21f10561)

This level of ringing is not acceptable in the maps, I suggest you try using a gaussian beam instead of a sigmoid as you were proposing in the last call

**Comment by fazlurahman** [comment]
> [@fazlurahman](https://github.com/fazlurahman) some recommendations on the notebooks:
> 
> * they should download all inputs instead of reading from local folders, e.g.
>   ```
>     gnilc_template = "varres"
>     dust_varresI = datadir / "COM_CompMap_Dust-GNILC-F353_2048_21p8acm.fits"
>     dust_varresP = (
>        datadir / f"COM_CompMap_IQU-thermaldust-gnilc-{gnilc_template}_2048_R3.00.fits"
>      )
>     if not dust_varresI.exists():
>     !wget -O $dust_varresI https://portal.nersc.gov/project/cmb/pysm-data/dust_gnilc/in
>   ```
> * if you use a data product which is not available publicly, please share it with me and I can put it under `pysm-data` at NERSC
> * they should save all temporary products to a `data/` subfolder
> * once you are happy with the results, do a "restart and run all" so you get a clean execution to share, this avoids unintentional out-of-order execution

@zonca, thanks for pointing this out. I will organize the notebooks as you suggested.

**Comment by fazlurahman** [comment]
> there should be something wrong, why the total Cls don't match with the large scale ones at low ells?

Yes @giuspugl, you are right. Something doesn't look right there. I need to take a closer look to figure out what’s happening there.

**Comment by fazlurahman** [comment]
On another note, I made a similar plot (with C_\ells at different stages) for the synchrotron template. Please see it below. It also has a similar trend (total Cls' is slightly higher than the only-large-scale Cls). 

![Image](https://github.com/user-attachments/assets/b5ee0627-6200-4601-b79d-48da648ccf9e)

**Comment by zonca** [comment]
@fazlurahman it would be good to have a cleaned up version of the notebook ready before the call tomorrow, so we can inspect it.

**Comment by fazlurahman** [comment]
Sure, I am on it.

**Comment by fazlurahman** [comment]
log_ss_modulated = log_ss * (
    hp.ud_grade(tsm 
                #/tsm.max()
                ,nside_out=hp.get_nside(log_ss))) 

This is a normalization done at the last step while modulating small-scales (tsm.max). When I commented it out, I see better agreement in the Cl's. 

![Image](https://github.com/user-attachments/assets/4a1704ba-ff7b-4d81-a6bf-95518fa115ac)

**Comment by fazlurahman** [comment]
And, the final map (ls+modulated*ss). Also, with Gaussian smoothing (instead of a sigmoid filter). 

![Image](https://github.com/user-attachments/assets/c11323ba-05c3-43f7-8abb-0b3559291d1b)

**Comment by fazlurahman** [comment]
I am trying to upload the notebook, but gist gives me some error. 

![Image](https://github.com/user-attachments/assets/1a9e1860-7151-4646-ac23-ca54f2a98197)

**Comment by zonca** [comment]
@fazlurahman try from the command line with `gh gist create --public` https://cli.github.com/

**Comment by fazlurahman** [comment]
Please find the notebook here.

https://gist.github.com/fazlurahman/c1e8b6a5e7f7af7fca448513e2becc60

Let me know if you have any questions.

**Comment by zonca** [comment]
@giuspugl what if we used a galactic mask? in any case we are interested in producing realistic small scales outside of the brightest regions of the Galaxy which are not usable for most CMB analysis anyway.

**Comment by fazlurahman** [comment]
Thanks @giuspugl for the suggestion. I will get back to it soon. 

I agree with @zonca’s point --  using a mask should help minimize ringing, especially for free-free emission (where ringing is mostly in the plane). In that case we should also use NaMaster for the C_\ell calculations (so far, I’ve been using anafast, as in the synchrotron notebook).

Related to this, I have a question: for computing C_\ell‘s on the circular patches, is it appropriate to use anafast? Since we’re working with a small region, I’m unsure if this is okay. It’s also possible that matrix inversion may not work reliably in NaMaster for such a small region, just a thought. 

@giuspugl, was this tested with both NaMaster and anafast in the synchrotron context? I see that the compute_cl_namaster function is defined in the notebook, but anafast was used.

**Comment by fazlurahman** [comment]
Related to this, I have a question: for computing C_\ell‘s on the circular patches, is it appropriate to use anafast? Since we’re working with a small region, I’m unsure if this is okay. It’s also possible that matrix inversion may not work reliably in NaMaster for such a small region, just a thought. 

@giuspugl, was this tested with both NaMaster and anafast in the synchrotron context? I see that the compute_cl_namaster function is defined in the notebook, but anafast was used.

**Comment by giuspugl** [comment]
> what if we used a galactic mask?
we indeed used a mask and didn't consider the small scales in the region of gal. plane, we directly employed the map as they are high SNR regions, 

> was this tested with both NaMaster and anafast in the synchrotron context?
Yes i did tests with both approaches .  problem with namaster is that we need the C_ells evaluated in one of the first bins ell =30-50 and that resulted in  unstable values (negative values or NaNs) , we agreed then to use anafast. I share your concerns @fazlurahman  but if window function effects are  affecting our estimates with anafast approach (and indeed they might ) they affect in the same way all the estimated spectra as the apodized mask  is the same . 
The changes in the C_ell   then should it be related somehow to galactic emission at large scales, that is what we roughly need to build the modulation maps. Of course please feel free to elaborate  something more clever or to explore better apodizations in order to have more stable spectra!

**Comment by fazlurahman** [comment]
Please find the modulation map here. 

![Image](https://github.com/user-attachments/assets/56ee0bb1-b3a2-4c56-8efd-600644091f47)

If this looks good to everyone, I will start focussing only on the ringing effects and small-scale addition. 

To me, the amplitude of the modulation map shows a much larger dynamic range than that of the synchrotron.

**Comment by zonca** [comment]
@fazlurahman any progress on this? now I have more time to look into this, can you please confirm if the latest version of the notebook is on the gist link you pasted before?

**Comment by zonca** [comment]
Thanks, the notebook now is running fine.
I am not sure I understand the benefit of the logpoltens formalism if there is no polarization.

**Comment by zonca** [comment]
@fazlurahman @giuspugl  do we need the logpoltens formalism given that there is no polarization?

**Comment by brandonshensley** [comment]
> Sorry [@zonca](https://github.com/zonca), I missed this question before.
> 
> One advantage I’ve seen with the log formalism is that it avoids the ringing we were getting when applying the sigmoid filter to the large-scale map.
> 
> Also, doing it in log space produce roughly lognormal maps in real space, which is a more realistic non-Gaussian model for the ISM.

I think this is exactly right: better behavior working in log space, plus pick up some non-Gaussianity. We don't have to call it log pol tensor, though I think it is equivalent with Q=U=0.

**Comment by zonca** [comment]
@fazlurahman use your own directory for now

**Comment by zonca** [comment]
see https://github.com/galsci/pysm/pull/256 this is ready for review by an expert

**Comment by fazlurahman** [comment]
I have prepared a notebook for the non-Gaussianity validation for the PySM free-free f2 model, based on Minkowski Functionals  and Skewness-Kurtosis statistics. 

Please go through it when you get a chance and let me know your thoughts.

https://gist.github.com/fazlurahman/2b75091ce019da5c19a6ac31de9f21e1
